onnxruntime/core/providers/shared_library/provider_interfaces.h

```diff
@@ -0,0 +1,480 @@
+  }
+
+  void operator=(const Provider_IAllocator& v) = delete;
+};
+
+template <size_t alignment>
+bool Provider_IAllocator::CalcMemSizeForArrayWithAlignment(size_t nmemb, size_t size, size_t* out) noexcept {
+  static constexpr size_t max_allowed = (static_cast<size_t>(1) << (static_cast<size_t>(std::numeric_limits<size_t>::digits >> 1))) - alignment;
+  static constexpr size_t max_size = std::numeric_limits<size_t>::max() - alignment;
+  static constexpr size_t alignment_mask = alignment - 1;
+  //Indeed, we only need to check if max_size / nmemb < size
+  //max_allowed is for avoiding unnecessary DIV.
+  if (nmemb >= max_allowed && max_size / nmemb < size) {
+    return false;
+  }
+  if (size >= max_allowed &&
+      nmemb > 0 && max_size / nmemb < size) {
+    return false;
+  }
+  if (alignment == 0)
+    *out = size * nmemb;
+  else
+    *out = (size * nmemb + alignment_mask) & ~static_cast<size_t>(alignment_mask);
+  return true;
+}
+
+struct Provider_IDeviceAllocator : Provider_IAllocator {
+  virtual bool AllowsArena() const = 0;
+};
+
+using Provider_AllocatorPtr = std::shared_ptr<Provider_IAllocator>;
+using Provider_DeviceAllocatorFactory = std::function<std::unique_ptr<Provider_IDeviceAllocator>(int)>;
+
+struct Provider_DeviceAllocatorRegistrationInfo {
+  OrtMemType mem_type;
+  Provider_DeviceAllocatorFactory factory;
+  size_t max_mem;
+};
+
+class TensorShape;
+
+struct Provider_Tensor {
+  virtual float* MutableData_float() = 0;
+  virtual const float* Data_float() const = 0;
+
+  template <typename T>
+  T* MutableData();
+
+  template <typename T>
+  const T* Data() const;
+
+  virtual const TensorShape& Shape() const = 0;
+};
+
+template <>
+inline float* Provider_Tensor::MutableData<float>() { return MutableData_float(); }
+
+template <>
+inline const float* Provider_Tensor::Data<float>() const { return Data_float(); }
+
+struct Provider_OpKernelInfo {
+  virtual Status GetAttr(const std::string& name, int64_t* value) const = 0;
+  virtual Status GetAttr(const std::string& name, float* value) const = 0;
+
+  template <typename T>
+  Status GetAttr(const std::string& name, T* value) const;
+};
+
+template <>
+inline Status Provider_OpKernelInfo::GetAttr<int64_t>(const std::string& name, int64_t* value) const {
+  return GetAttr(name, value);
+}
+
+template <>
+inline Status Provider_OpKernelInfo::GetAttr<float>(const std::string& name, float* value) const {
+  return GetAttr(name, value);
+}
+
+struct Provider_OpKernelContext {
+  virtual const Provider_Tensor* Input_Tensor(int index) const = 0;
+
+  template <typename T>
+  const T* Input(int index) const;
+
+  virtual Provider_Tensor* Output(int index, const TensorShape& shape) = 0;
+};
+
+template <>
+inline const Provider_Tensor* Provider_OpKernelContext::Input<Provider_Tensor>(int index) const {
+  return Input_Tensor(index);
+}
+
+struct Provider_OpKernel {
+  Provider_OpKernel(const Provider_OpKernelInfo& /*info*/) {}
+  virtual ~Provider_OpKernel() = default;
```

Comment:
Should the base struct store the 'info'?

Reply:
It was only needed by the actual OpKernel base class. The wrapper never accesses it.